### PR TITLE
NT-19 effect calculation

### PR DESCRIPTION
**Pull request by PatrickSlag**
Added a part to the algorithm to check the effect if the originator of the trade builds the structure.

### CODE REVIEW

File: Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+using Models;
+
+namespace LogicServices.Algorithm
+{
+    public class BuildEffect
+    {
+        public bool Calculate(Trade trade, User target, User originator)
+        {
+            var points = target.PointTable[(trade.RequestedItem, originator)];
+            return points switch
```
Comment (jortdr) [inline Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs:24]: would love a comment here to explain what it does

**Review by jortdr — DISMISSED** [review]

File: Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs (hunk 1 above)
Comment (VarioCode) [reply Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs:24]: What a lovely idea @PatrickSlag would be real nice

File: Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+            {
+                5 => true,
+                0 => true,
```
Comment (VarioCode) [inline Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs]: Shouldn't this be a "coin flip"?

File: Assets/Tests/AlgorithmTests.cs [hunk 3]
```diff
@@ -61,16 +61,82 @@ public void CalculateSelfBuild_LowerThanBorder_ReturnsTrue()
...
+            
+            //When
+            target.PointTable = new Dictionary<(InventoryItems, User), int>
+            {
+                [(InventoryItems.Wood, target)] = 10,
+                [(InventoryItems.Wood, originator)] = -5
+            };
+            
+            //Then
+            Assert.IsFalse(buildEffect.Calculate(trade,target,originator));
```
Comment (VarioCode) [inline Assets/Tests/AlgorithmTests.cs]: Do not use methods in asserts when writing tests

File: Assets/Tests/AlgorithmTests.cs [hunk 4]
```diff
@@ -61,16 +61,82 @@ public void CalculateSelfBuild_LowerThanBorder_ReturnsTrue()
...
+            
+            //When
+            target.PointTable = new Dictionary<(InventoryItems, User), int>
+            {
+                [(InventoryItems.Wood, target)] = 10,
+                [(InventoryItems.Wood, originator)] = 0
+            };
+            
+            //Then
+            Assert.IsTrue(buildEffect.Calculate(trade,target,originator));
```
Comment (VarioCode) [inline Assets/Tests/AlgorithmTests.cs]: Do not use methods in asserts when writing tests

File: Assets/Tests/AlgorithmTests.cs [hunk 5]
```diff
@@ -61,16 +61,82 @@ public void CalculateSelfBuild_LowerThanBorder_ReturnsTrue()
...
+            
+            //When
+            target.PointTable = new Dictionary<(InventoryItems, User), int>
+            {
+                [(InventoryItems.Wood, target)] = 10,
+                [(InventoryItems.Wood, originator)] = 5
+            };
+            
+            //Then
+            Assert.IsTrue(buildEffect.Calculate(trade,target,originator));
```
Comment (VarioCode) [inline Assets/Tests/AlgorithmTests.cs]: Do not use methods in asserts when writing tests

File: Assets/Scripts/LogicServices/AlgorithmService.cs [hunk 6]
```diff
@@ -29,8 +31,15 @@ public bool Decide(Trade trade,User originator, User targetCpu)
...
-            if (_randomness.Calculate()) return !_selfBuild.Calculate(trade,targetCpu);
-            return _selfBuild.Calculate(trade,targetCpu);
+            //Decisions
+            bool randomDecision = _randomness.Calculate();
+            bool selfBuildDecision = _selfBuild.Calculate(trade, targetCpu);
+            bool buildEffectDecision = _buildEffect.Calculate(trade, targetCpu, originator);
+            
+            Debug.Log($"Random: {randomDecision} SelfBuild: {selfBuildDecision} BuildEffect: {buildEffectDecision}");
+
+            if (randomDecision) return !(selfBuildDecision && buildEffectDecision);
```
Comment (VarioCode) [inline Assets/Scripts/LogicServices/AlgorithmService.cs]: The Randomness should be applied to a single decision, not to the whole group of decisions.

**Review by VarioCode — CHANGES_REQUESTED** [review]
Used methods in asserts again, also added a comment to the use of randomness in this PR.

File: Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs (hunk 2 above)
Comment (jortdr) [reply Assets/Scripts/LogicServices/Algorithm/BuildEffect.cs]: 0 should indeed be a coin flip

File: Assets/Scripts/LogicServices/AlgorithmService.cs [hunk 7]
```diff
@@ -11,18 +12,29 @@ public class AlgorithmService
...
         public bool Decide(Trade trade,Tribe originator, Tribe targetCpu)
         {
-            if (_randomness.Calculate()) return !_selfBuild.Calculate(trade,targetCpu);
-            return _selfBuild.Calculate(trade,targetCpu);
+            //Original Decisions
+            bool selfBuildDecision = _selfBuild.Calculate(trade, targetCpu);
+            bool buildEffectDecision = _buildEffect.Calculate(trade, targetCpu, originator);
+            
+            //Randomise Decisions
+            selfBuildDecision = _randomness.Calculate() ? selfBuildDecision : !selfBuildDecision;
```
Comment (VarioCode) [inline Assets/Scripts/LogicServices/AlgorithmService.cs]: Why is the "randomness chance" the same? Shouldn't it be configurable based on the decision it used for?

**Review by VarioCode — CHANGES_REQUESTED** [review]

File: Assets/Scripts/LogicServices/AlgorithmService.cs [hunk 8]
```diff
@@ -1,6 +1,7 @@
-using System;
 using LogicServices.Algorithm;
 using Models;
+using UnityEngine;
```
Comment (VarioCode) [inline Assets/Scripts/LogicServices/AlgorithmService.cs]: Don't use UnityEngine in the service layer

**Review by VarioCode — CHANGES_REQUESTED** [review]

File: Assets/Scripts/LogicServices/AlgorithmService.cs [hunk 9]
```diff
@@ -11,18 +11,27 @@ public class AlgorithmService
...
         {
-            if (_randomness.Calculate()) return !_selfBuild.Calculate(trade,targetCpu);
-            return _selfBuild.Calculate(trade,targetCpu);
+            //Original Decisions
+            bool selfBuildDecision = _selfBuild.Calculate(trade, targetCpu);
+            bool buildEffectDecision = _buildEffect.Calculate(trade, targetCpu, originator);
+            
+            //Randomise Decisions
+            selfBuildDecision = _randomness.Calculate(0.2f) ? selfBuildDecision : !selfBuildDecision;
+            buildEffectDecision = _randomness.Calculate(0.1f) ? buildEffectDecision : !buildEffectDecision;
```
Comment (jortdr) [inline Assets/Scripts/LogicServices/AlgorithmService.cs]: I'd probably prefer to have these randomness variables somewhere up top for easy access and tweaking, perhaps possibly even in editor.

**Review by jortdr — COMMENTED** [review]
lgtm

File: Assets/Scripts/LogicServices/AlgorithmService.cs (hunk 9 above)
Comment (VarioCode) [reply Assets/Scripts/LogicServices/AlgorithmService.cs]: Yea, I agree, I would also prefer to use a variable instead of these magic numbers.

**Review by VarioCode — APPROVED** [review]

**Review by jortdr — APPROVED** [review]

**Review by VarioCode — APPROVED** [review]

**Pull request merged**